reference/src/client/test/sa_key_provision_applefairplay.cpp

```diff
@@ -0,0 +1,284 @@
+           appleFPProvision, 0, nullptr);
+        if (nullptr != appleFPProvision) {
+           delete appleFPProvision;
+           appleFPProvision = nullptr;
+        }
+        ASSERT_NE(status, SA_STATUS_OK);
+    }
+    TEST_F(SaKeyProvisionAppleFairplayTest, failNullProvision) {
+        auto key_length = SYM_128_KEY_SIZE;
+        std::vector<uint8_t> clear_secret_key = random(key_length);
+
+        //create a base key container
+        auto secret_key_type = SA_KEY_TYPE_SYMMETRIC;
+        std::string secret_key_type_string = "AES-128";
+
+        std::string secret_key_container;
+        sa_status status = create_key_container(
+             secret_key_type_string, //key type string
+             secret_key_type, //key type
+             clear_secret_key,
+             secret_key_container);
+        ASSERT_EQ(SA_STATUS_OK, status);
+        INFO("secret_key_container length : %d", secret_key_container.size());
+
+        AppleFairPlayProvisioning *appleFPProvision = new AppleFairPlayProvisioning;
+        ASSERT_NE(nullptr, appleFPProvision);
+        //Note: here just provide an examlpe to create a secret key and passed
+        //into sa_key_provision_ta, how to use it is all up to SOC vendors.
+        appleFPProvision->fairPlaySecret = (void*)secret_key_container.data();
+        appleFPProvision->fairPlaySecretLength = secret_key_container.size();
+
+        status = sa_key_provision_ta(APPLE_FAIRPLAY_PROVISIONING, nullptr,
+           sizeof(AppleFairPlayProvisioning), nullptr);
+
+        if (nullptr != appleFPProvision) {
+           delete appleFPProvision;
+           appleFPProvision = nullptr;
+        }
+        ASSERT_NE(status, SA_STATUS_OK);
+    }
+#ifdef FILE_BASED_FETCH_KEY
+    //There is a file  apple_fairplay_secret.key under tasecureapi/reference/src/client/,
+    //this file is a fairplay secret key. you can do
+    //"export  apple_fairplay_secret_key=~/PATH/tasecureapi/reference/src/client/apple_fairplay_secret.key",
+    //the following test fromFileBased will pick up them and test
+    //Or
+    //you just simply copy these two files and put under /opt/drm/
+ TEST_F(SaKeyProvisionAppleFairplayTest, fromFileBased) {
+        AppleFairPlayProvisioning *appleFPProvision = new AppleFairPlayProvisioning;
+        ASSERT_NE(nullptr, appleFPProvision);
+        ASSERT_TRUE(readAppleFairplayData(&appleFPProvision));
+
+        sa_status status = sa_key_provision_ta(APPLE_FAIRPLAY_PROVISIONING,
+          appleFPProvision, sizeof(appleFPProvision), nullptr);
+        ASSERT_EQ(status, SA_STATUS_OK);
+
+        if (nullptr != appleFPProvision->fairPlaySecret){
+           free(appleFPProvision->fairPlaySecret);
+           appleFPProvision->fairPlaySecret = nullptr;
+        }
+    }
+#endif //FILE_BASED_FETCH_KEY
+} // namespace
+
+#ifdef FILE_BASED_FETCH_KEY
+#include <sys/stat.h>
+#include <string.h>
+
+#define apple_fairplay_secret_key "/opt/drm/apple_fairplay_secret.key"
```

Comment:
we shouldn't use the /opt/drm folder as it is an implementation specific location.